src/types.ts

```diff
@@ -616,6 +617,10 @@ const ipv6Regex =
 const base64Regex =
   /^([0-9a-zA-Z+/]{4})*(([0-9a-zA-Z+/]{2}==)|([0-9a-zA-Z+/]{3}=))?$/;
 
+// https://base64.guru/standards/base64url
+const base64urlRegex =
+  /^([0-9a-zA-Z-_]{4})*(([0-9a-zA-Z-_]{2}(==)?)|([0-9a-zA-Z-_]{3}(=)?))?$/;
```

Comment:
According to the spec, excess pad characters may be ignored:
https://datatracker.ietf.org/doc/html/rfc4648#section-3.3
Why not use the simpler regex `/^[0-9a-zA-Z-_=]*$/`?

Reply:
This regex would allow `=` at any position in the string, which is not valid `base64url`. Even a regex allowing `=` only at its end would allow invalid strings like `ABCDE`; here, the characters `ABCD` encode three bytes of information, and the `E` does not hold enough information for a full octet. The corresponding `base64` expression would be `ABCDE===` with three padding characters, which is not allowed either.

The existing `base64Regex` is rather advanced and validates things like “only multiples of four characters allowed” and “at most two padding characters allowed”. As `base64url` is derived from `base64` with the addition “padding can be skipped”, my `base64urlRegex` is derived from `base64Regex` by making padding characters optional, which retains the validations for the aforementioned edge cases.

Reply:
> This regex would allow = at any position in the string, which is not valid base64url

By my reading of section 3.3, it is valid. The spec allows extra padding as valid input to a decoder:

> Furthermore, such specifications MAY ignore the pad character, "=", treating it as non-alphabet data, if it is present before the end of the encoded data.  If more than the allowed number of pad characters is found at the end of the string (e.g., a base 64 string terminated with "==="), the excess pad characters MAY also be ignored.

On the other hand, by section 3.2, an encoder *must* include the padding:

> Implementations MUST include appropriate pad characters at the end of encoded data unless the specification referring to this document explicitly states otherwise.

So I guess the question is should this err on the side of leniency or strictness?

Reply:
Okay, I didn’t catch that after my first reading of the spec. But if I understand it correctly, the “specifications” in

> Furthermore, such specifications MAY ignore the pad character, "=", …

refer to the sentence at the beginning of the paragraph

> …, unless the specification referring to this document explicitly states otherwise.

meaning that a specification based on base64 like MIME or PEM can decide to ignore pad characters before the end of the encoded data, just as MIME ignores CRLFs instead of rejecting input containing those non-alphabet characters. Section 5 of the RFC, specifying base64url, neither explicitly states that non-alphabet characters or padding characters before the end of the encoded data may be ignored (it only allows “skipping the padding” while referring to section 3.2, which described padding at the end of the encoded data), nor does it explicitly reference section 3.3. In my understanding, this would not allow padding characters before the end of the encoded data.

Also, JavaScript runtimes like Node.js or Bun do not simply ignore those padding characters and instead terminate decoding when reaching them:

```js
// Node.js:
> Buffer.from("SGVsbG8gV29ybGQ=", "base64url")
<Buffer 48 65 6c 6c 6f 20 57 6f 72 6c 64>
> Buffer.from("SGVsbG8gV29ybGQ", "base64url")
<Buffer 48 65 6c 6c 6f 20 57 6f 72 6c 64>
> Buffer.from("SGVs=bG8gV29ybGQ", "base64url")
<Buffer 48 65 6c>

// Bun:
> Buffer.from("SGVsbG8gV29ybGQ=", "base64url")
Buffer(11) [Uint8Array] [
   72, 101, 108, 108,
  111,  32,  87, 111,
  114, 108, 100
]
> Buffer.from("SGVsbG8gV29ybGQ", "base64url")
Buffer(11) [Uint8Array] [
   72, 101, 108, 108,
  111,  32,  87, 111,
  114, 108, 100
]
> Buffer.from("SGVs=bG8gV29ybGQ", "base64url")
Buffer(3) [Uint8Array] [ 72, 101, 108 ]
```

So perhaps I am wrong in interpreting the specification, but if so, runtimes aren’t better at that.

Reply:
> Also, JavaScript runtimes like Node.js or Bun do not simply ignore those padding characters and instead terminate decoding when reaching them

Oof! That's an excellent observation and a great reason to reject `=` in the middle! Also it seems like the trailing `=` is not handled particularly well, so I think your regex hits the perfect balance.